src/main/java/com/allenai/ml/optimize/QuasiNewton.java

```diff
@@ -0,0 +1,96 @@
+package com.allenai.ml.optimize;
+
+import com.allenai.ml.linalg.Vector;
+import com.gs.collections.api.tuple.Pair;
+import com.gs.collections.impl.tuple.Tuples;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Interface for how to approximate `H^{-1} dir`, where `H^{-1}` is an approximation for the inverse hessian and
+ * dir is the current search direction. If we use `H^{-1}` as the identity matrix, you get standard gradient descent.
+ * If you remember some history, you get LBFGS.
+ */
+@FunctionalInterface
+public interface QuasiNewton {
+
+    /**
+     * @return Approximation of inverse hessian matrix times `dir`
+     */
+    Vector implictMultiply(Vector dir);
+
+    /**
+     * Update the approximation
+     * @param xDelta The delta from the last guess
+     * @param gradDelta The delta from the function gradient at last guess
+     */
+    default void update(Vector xDelta, Vector gradDelta) {
+        // intentional no-op
+    }
+
+    /**
+     * Approximate inverse hessian with the identity
+     */
+    static QuasiNewton gradientDescent() {
+        return dir -> dir;
+    }
+
+    /**
+     * LBFGS inverse hessian approximation
+     */
+    static QuasiNewton lbfgs(final int maxHistorySize) {
+        return new QuasiNewton() {
+            // Store (xDelta, gradDelta) pairs
+            private final List<Pair<Vector, Vector>> history = new ArrayList<>();
+
+            private double initialScale() {
+                if (history.isEmpty()) {
+                    return 1.0;
+                }
+                Vector lastInputDiff = history.get(0).getOne();
+                Vector lastGradDiff = history.get(0).getTwo();
+                double numer = lastGradDiff.dotProduct(lastInputDiff);
+                double denom = lastGradDiff.l2NormSquared();
+                if (denom == 0.0) {
+                    throw new RuntimeException("Can't have zero curvature");
+                }
+                return numer / denom;
+            }
+
+            @Override
+            public Vector implictMultiply(Vector dir) {
+                double[] rho = new double[history.size()];
+                double[] alpha = new double[history.size()];
+                Vector right = dir.copy();
+                for (int i = history.size() - 1; i >= 0; i--) {
+                    Vector inputDifference = history.get(i).getOne();
+                    Vector derivativeDifference = history.get(i).getTwo();
+                    rho[i] = inputDifference.dotProduct(derivativeDifference);
+                    if (rho[i] == 0.0) {
+                        throw new RuntimeException("LBFGSMinimizer.implicitMultiply: Curvature problem.");
```

Comment:
Similar to the above. The assumption is that you're actually taking steps in the overall optimization code (there is a non-zero l2 delta between successive inputs); when that doesn't happen, some of the mathematical terms needed aren't defined.

This too should maybe be an `IllegalArgumentException` or something that indicates you did something wrong above me.